src/marqo/core/inference/device_manager.py

```
@@ -0,0 +1,94 @@
+        """
+        if not self._is_cuda_available_at_startup:
+            # If the instance is initialised without cuda devices, skip the check
+            return
+
+        if not torch.cuda.is_available():
+            # CUDA devices could become unavailable/unreachable if the docker container running Marqo loses access
+            # to the device symlinks. There is no way to recover from this, we will need to restart the container.
+            # See https://github.com/NVIDIA/nvidia-container-toolkit/issues/48 for more details.
+            logger.error('Cuda device becomes unavailable')
+            raise CudaDeviceNotAvailableError('Cuda device becomes unavailable')
+
+        # TODO confirm whether we should check all devices or just the default one
+        for device in self.cuda_devices:
+            try:
+                cuda_device = torch.device(device.name)
+                memory_stats = torch.cuda.memory_stats(cuda_device)
+                logger.debug(f'Cuda device {device.name} with total memory {device.total_memory}. '
+                             f'Memory stats: {str(memory_stats)}')
+
+                torch.randn(3, device=cuda_device)
+            except RuntimeError as e:
+                if 'out of memory' in str(e).lower():
+                    # If we encounter 'CUDA error: out of memory' error consistently, it means some threads are
+                    # holding the memory
+                    logger.error(f'Cuda device {device.name} is out of memory. Total memory: {device.total_memory}. '
+                                 f'Memory stats: {str(memory_stats)}')
+                    allocated_mem = memory_stats.get("allocated.all.current", None) if memory_stats else None
+                    raise CudaOutOfMemoryError(f'Cuda device {device.name} is out of memory: '
+                                               f'({allocated_mem}/{device.total_memory})')
+            except Exception as e:
```

This exception catch seems too broad. This may mask CUDA issues in the future

---

The purpose is to catch all CUDA errors or non-CUDA related errors, so that we don't crash the container on unknown errors.

---

I agree catching Exception is dubious. If there's a bug in the code you'll just mask it and log a warning. An error here won't crash the container, it'll cause a 500 API response which is fine for an unhandled situation

---

I agree we should still raise a 500 if other errors occur aside from oom

---

>  An error here won't crash the container, it'll cause a 500 API response which is fine for an unhandled situation
If this endpoint is used by scheduler, we will need to be cautious. A error might be recoverable or transient. If we restart the container on these errors, we might introduce unnecessary downtime. (Please note if critical error like OOM or cuda unavailable happens, there's already downtime, so it's ok for rescheduler to shut it down and reschedule).

---

That's a fair point, as long as cloud team consistently looks at these warnings to make sure no we're not silently failing often